solarkraft/src/state/value.ts

```diff
@@ -137,14 +183,84 @@ export function addr(s: string): Value {
 
 // Byte arrays (Bytes, BytesN)
 // The `len` field is present iff the length is fixed (i.e. for BytesN)
-// TODO: tests
 export type ArrValue = {
-    val: { type: "u32", val: bigint }[]
+    val: u32T[]
     type: "arr"
     len?: number
-    // if (typeof (l) !== 'undefined' && v.length !== l) {
-    //     throw new TypeError(`Array declared as fixed-length ${l}, but actual length is ${v.length}.`)
-    // }
+}
+
+// Safe constructor for arr-typed `Value`s representing non-fixed width byte arrays. 
+// Throws a `TypeError` if `v` contains an invalid value.
+export function bytes(v: u32T[]): Value {
+    const obj: ArrValue = { type: "arr", val: v }
+    if (!isValid(obj)) {
+        throw new TypeError(`Some element of ${v} is not valid.`)
+    }
+    return obj
+}
+
+// Safe constructor for arr-typed `Value`s representing fixed width byte arrays. 
+// Throws a `TypeError` if `v` contains an invalid value.
+export function bytesN(v: u32T[]): ArrValue {
+    const obj: ArrValue = { type: "arr", val: v, len: v.length }
+    if (!isValid(obj)) {
+        throw new TypeError(`Some element of ${v} is not valid.`)
+    }
+    return obj
+}
+
+// Vectors are an ordered collection of `Value`s, with possible duplicates.
+// The values in a Vec are not guaranteed to be of any specific type. <-- from the docs
+export type VecValue = {
+    val: Value[]
+    type: "vec"
+}
+
+// Safe constructor for vec-typed `Value`s. Throws a `TypeError` if `v` contains an invalid value.
+export function vec(v: Value[]): VecValue {
+    const obj: VecValue = { type: "vec", val: v }
+    if (!isValid(obj)) {
+        throw new TypeError(`Some element of ${v} is not valid.`)
+    }
+    return obj
+}
+
+// Soroban Map is an ordered key-value dictionary (note that JS maps are in principle unordered, but will iterate in insertion order).
```

Comment:
Would it make sense to use [OrderedMap](https://immutable-js.com/docs/v3.8.2/OrderedMap/) here? We should also make the decision whether we want our values immutable or mutable. I would prefer immutable for our application.